takashi-task-framework.md

```diff
@@ -0,0 +1,265 @@
+dependency checks.
+
+Let's say we want to make sure that `pandoc` is on the `$PATH`, because
+a task we want to describe requires pandoc.
+
+// check-pandoc.sh
+
+```sh
+#!/usr/bin/env bash
+
+which pandoc # returns a status code of 0 if found, 1 if not.
+```
+
+We can see that this command will return a `0` if `pandoc` is on the
+`$PATH`, or a `1` if not.
+
+Thus, we can compose our dependencies by having our task description run
+bash scripts in the order they are placed in the config file, until they
+all return 0 or one script returns a non-zero exit code. In case a
+script returns a non-zero exit code, we can bubble up the error to the
+user (clearly this task can't run).
+
+Let's assume we'll use some key-value format, like an INI format.
+
+```ini
+[Checks]
+check-pandoc="./check-pandoc"
+```
+
+Next, we want to make sure that we can pass environment variables, and
+command line arguments to our script:
+
+```ini
+[Checks]
+check-pandoc="./check-pandoc"
+
+[Environment]
+DOCKER_PORT="0.0.0.0"
+
+[Args]
+args="site.md -o site.html --template=_template.html"
+
+[Command]
+command="pandoc"
+```
+
+Finally, we need to find some way to express dependencies:
+
+```ini
+[Dependencies]
+Requires=libpng.service # make sure that libpng is running.
+After=network.target # make sure the network is running
+
+[Service]
+User=test
+Group=test
+WorkingDirectory=/tmp/pandoc
+Restart=never
+LimitMem=infinity # infinite memory
+MaxTime=2s # if it doesn't return in 2s, kill this job
+
+[Checks]
+check-pandoc="./check-pandoc"
+
+[Environment]
+DOCKER_PORT="0.0.0.0"
+
+[Args]
+args="site.md -o site.html --template=_template.html"
+
+[Command]
+command="pandoc"
+```
+
+Where libpng.service might look like this:
+
+```ini
+[Dependencies]
+Requires=libjpeg.service # make sure that libjpeg is running
+```
+
+So we can express dependencies of dependencies. We'll resolve
+dependencies of dependencies by doing a topological sort, where we start
+at the top of the dependency list and make sure they all successfully
+run before going to the task at hand.
+
+## Running Jobs
+
+We'll also want a way to convey a task should be run periodically or
+after something occurs: (e.g. an Account has too many inactive users
+which need to be pruned for better performance on hot storage).
+
+In the case of periodic jobs, we can support cron-like syntax in our INI
+config files to allow for a task to fire every so often.
+
+In the case of a condition being met, we'll allow the user to configure
+how often they want to check for a particular condition and the action
+they wish to do:
+
+```ini
+[RunIf]
+Requires="./check-if-too-many-users.sh" # checks for user count of all
+accounts, returns 0 if the task needs to run, 1 otherwise.
+
+[Command]
+command="./prune-users.sh"
+```
+
+Otherwise we can support a cron-like syntax:
+
+```ini
+[RunIf]
+cron="* * * * *" # run every minute
+
+[Command]
+command="./heartbeat.sh" # check if the other server is still up.
+```
```

Comment:
you talk about this idea of shell scripts or bash scripts used to enforce a certain configuration, why is it preferably done this way? if we allow users of the framework to write their own scripts for montioring and such how do we ensure across teams that we aren't rewriting code? What can we do to maybe make this process more standardized?

Reply:
This is a tradeoff. Using shell scripts versus something more structured like dhall depends on the team -- some people prefer shell scripts and some people prefer more structure.

To deal with the fragmentation argument, we could have a scripts team that exposes common script functionality and documents recipes for engineers to use. At many companies I've worked for, this is one way to decrease duplication, as some teams will be incentivized to maintain boilerplate repos/libraries for internal users.